packages/@aws-cdk/aws-rds/lib/cluster-engine.ts

```diff
@@ -0,0 +1,286 @@
+   *
+   * @param engineVersion the exact engine version, for example "5.7.mysql_aurora.2.03.4"
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html#cfn-rds-dbcluster-engineversion
+   */
+  withVersion(engineVersion: string): IClusterEngine;
+
+  /**
+   * Method called when the engine is used to create a new cluster.
+   */
+  bindToCluster(scope: core.Construct, options: ClusterEngineBindOptions): ClusterEngineBindConfig;
+}
+
+abstract class AbstractClusterEngine implements IClusterEngine {
+  public readonly engineType: string;
+  public readonly engineVersion?: string;
+  public readonly parameterGroupFamily: string;
+  public readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
+  public readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
+  // this works for all 3 cluster engines that are currently supported
+  public readonly defaultInstanceType = ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MEDIUM);
+
+  private readonly needsS3RolesInParameters: boolean;
+  private readonly parameterGroupFamilies?: ParameterGroupFamilyMapping[];
+  private readonly defaultPort?: number;
+
+  constructor(
+    engineName: string,
+    needsS3RolesInParameters: boolean,
+    singleUserRotationApplication: secretsmanager.SecretRotationApplication,
+    multiUserRotationApplication: secretsmanager.SecretRotationApplication,
+    parameterGroupFamilies?: ParameterGroupFamilyMapping[],
+    defaultPort?: number,
+    engineVersion?: string) {
+
+    this.engineType = engineName;
+    this.needsS3RolesInParameters = needsS3RolesInParameters;
+    this.singleUserRotationApplication = singleUserRotationApplication;
+    this.multiUserRotationApplication = multiUserRotationApplication;
+    this.parameterGroupFamilies = parameterGroupFamilies;
+    this.defaultPort = defaultPort;
+    this.engineVersion = engineVersion;
+    this.parameterGroupFamily = this.establishParameterGroupFamily();
+  }
+
+  public withVersion(engineVersion: string): IClusterEngine {
+    const self = this;
+    // put it in a local variable, as it's protected
+    const defaultParameterGroup = this.defaultParameterGroup;
+
+    class CopyWithVersion extends AbstractClusterEngine {
+      protected defaultParameterGroup(scope: core.Construct) {
+        // we need to use call, so that 'this' inside defaultParameterGroup() is resolved correctly
+        return defaultParameterGroup.call(self, scope);
+      }
+    }
+
+    return new CopyWithVersion(
+      this.engineType,
+      this.needsS3RolesInParameters,
+      this.singleUserRotationApplication,
+      this.multiUserRotationApplication,
+      this.parameterGroupFamilies,
+      this.defaultPort,
+      engineVersion,
+    );
+  }
+
+  public bindToCluster(scope: core.Construct, options: ClusterEngineBindOptions): ClusterEngineBindConfig {
```

Comment:
1. `bindToCluster()` should take the cluster as a parameter, even if it's not used in the current implementations. It's reasonable to expect both operands in an operation's implementation.

2. Instead of `ClusterEngineBindOptions`, can we instead configure the `s3ImportRole` and `s3ExportRole` as constructor arguments of the Aurora implementation of `IClusterEngine`? I believe this is a feature supported only by Aurora. It may be valuable to have a mechanism to configure/pass parameters that are unique to the engine?

Reply:
1. Funny enough, I had it take `IDatabaseCluster` at the beginning. But the problem is that it doesn't really work. The problem is that `bindToCluster()` has to be called in the constructor of `DatabaseCluster`. Which means you'll actually pass a partially-constructed object to `bindToCluster()`, which will most likely result in errors when accessing any of its properties.

2. No. This is supported by 2 engines ("vanilla" Aurora, and Aurora MySQL).